src/b-splines/cubic.jl

```diff
@@ -0,0 +1,241 @@
+        $symixp = mod1($symix + 1, size(itp,$d))
+        $symixpp = mod1($symix + 2, size(itp,$d))
+    end
+end
+
+# TODO: implement versions of `define_indices_d` for different Bc's if necessary
+
+padding{BC<:Flag}(::Type{BSpline{Cubic{BC}}}) = Val{1}()
+padding(::Type{BSpline{Cubic{Periodic}}}) = Val{0}()
+
+"""
+In this function we assume that `fx_d = x-ix_d` and we produce `cX_d` for
+`X ⋹ {m, _, p, pp}` such tha
+
+    cm_d  = p(fx_d)
+    c_d   = q(fx_d)
+    cp_d  = q(1-fx_d)
+    cpp_d = p(1-fx_d)
+
+where `p` and `q` are defined in the docstring entry for `Cubic`.
+"""
+function coefficients{C<:Cubic}(::Type{BSpline{C}}, N, d)
+    symm, sym =  symbol("cm_",d), symbol("c_",d)
+    symp, sympp = symbol("cp_",d) ,symbol("cpp_",d)
+    symfx = symbol("fx_",d)
+    symfx_cub = symbol("fx_cub_", d)
+    sym_1m_fx_cub = symbol("one_m_fx_cub_", d)
+    quote
+        $symfx_cub = cub($symfx)
+        $sym_1m_fx_cub = cub(1-$symfx)
+        $symm = SimpleRatio(1,6)*$sym_1m_fx_cub
+        $sym  = SimpleRatio(2,3) - sqr($symfx) + SimpleRatio(1,2)*$symfx_cub
+        $symp = SimpleRatio(2,3) - sqr(1-$symfx) + SimpleRatio(1,2)*$sym_1m_fx_cub
+        $sympp = SimpleRatio(1,6)*$symfx_cub
+    end
+end
+
+# This assumes integral values ixm_d, ix_d, and ixp_d,
+# coefficients cm_d, c_d, and cp_d, and an array itp.coefs
+function index_gen{C<:Cubic,IT<:DimSpec{BSpline}}(::Type{BSpline{C}}, ::Type{IT}, N::Integer, offsets...)
+    if length(offsets) < N
+        d = length(offsets)+1
+        symm, sym, symp, sympp =  symbol("cm_",d), symbol("c_",d), symbol("cp_",d), symbol("cpp_",d)
+        return :($symm * $(index_gen(IT, N, offsets...,-1)) + $sym * $(index_gen(IT, N, offsets..., 0)) +
+                 $symp * $(index_gen(IT, N, offsets..., 1)) + $sympp * $(index_gen(IT, N, offsets..., 2)))
+    else
+        indices = [offsetsym(offsets[d], d) for d = 1:N]
+        return :(itp.coefs[$(indices...)])
+    end
+end
+
+# ------------ #
+# Prefiltering #
+# ------------ #
+
+function inner_system_diags{T,C<:Cubic}(::Type{T}, n::Int, ::Type{C})
+    du = fill(convert(T, SimpleRatio(1, 6)), n-1)
+    d = fill(convert(T, SimpleRatio(2, 3)), n)
+    dl = copy(du)
+    dl, d, du
+end
+
+"""
+`Flat` `OnGrid` amounts to setting `y_0'(x) = 0` at `x = 0`. Applying this gives
+a condition where:
+
+    -cm + cp = 0
+"""
+function prefiltering_system{T,TC}(::Type{T}, ::Type{TC}, n::Int,
+                                   ::Type{Cubic{Flat}}, ::Type{OnGrid})
+    dl, d, du = inner_system_diags(T, n, Cubic{Flat})
+    d[1] = d[end] = -one(T)  # set first and last diagonal to -1
+    du[1] = dl[end] = zero(T)  # set just off diagonal to 0
+
+    # Now Woodbury correction to set `[1, 3], [n, n-2] ==> 1`
+    specs = _build_woodbury_specs(T, n, (1, 3, one(T)), (n, n-2, one(T)))
+
+    Woodbury(lufact!(Tridiagonal(dl, d, du), Val{false}), specs...), zeros(TC, n)
+end
+
+# TODO: check to see if above method is the same as Quadratic
```

Comment:
The mathematical root would of course be the same - `∇f(x_0) = 0` for `x_0` at the boundary - however that doesn't necessarily mean the same thing for the coefficient equations. I *think* we need to handle cubic separately.

If you work out the equations and find that they are indeed the same as for the quadratic case, I would prefer a method here that forwards to the quadratic (to make the re-use explicit in this file) rather than just updating the parameter types on the quadratic method.